File: go/parsers/welcome.go
package parsers

import (
	"capnproto.org/go/capnp/v3"

	"github.com/xconnio/wampproto-go/messages"
	"github.com/xconnio/wampproto-serializer-capnproto/go/gen"
)

type Welcome struct {
	gen *gen.Welcome
}

func NewWelcomeFields(g *gen.Welcome) messages.WelcomeFields {
	return &Welcome{gen: g}
}

func (w *Welcome) SessionID() uint64 {
	return w.gen.SessionID()
}

func (w *Welcome) Details() map[string]any {
	authID, _ := w.gen.Authid()
	authRole, _ := w.gen.Authrole()
	authMethod, _ := w.gen.Authmethod()
	authProvider, _ := w.gen.Authprovider()

	details := map[string]any{
		"authid":       authID,
		"authrole":     authRole,
		"authmethod":   authMethod,
		"authprovider": authProvider,
	}

	genRoles, err := w.gen.Roles()
	if err != nil {
		return details
	}

	roles := map[string]any{}
	if r, err := genRoles.Dealer(); err == nil {
		dealer := map[string]any{}
		if r.CallerIdentification() {
			dealer["caller_identification"] = true
		}
		if r.CallTimeout() {
			dealer["call_timeout"] = true
		}
		if r.CallCanceling() {
			dealer["call_canceling"] = true
		}
		if r.ProgressiveCallResults() {
			dealer["progressive_call_results"] = true
		}
		if r.PatternBasedRegistration() {
			dealer["pattern_based_registration"] = true
		}
		if r.SharedRegistration() {
			dealer["shared_registration"] = true
		}
		roles["dealer"] = dealer
	}

	if r, err := genRoles.Broker(); err == nil {
		broker := map[string]any{}
		if r.PublisherIdentification() {
			broker["publisher_identification"] = true
		}
		if r.PublisherExclusion() {
			broker["publisher_exclusion"] = true
		}
		if r.AcknowledgeEventReceived() {
			broker["acknowledge_event_received"] = true
		}
		if r.PatternBasedSubscription() {
			broker["pattern_based_subscription"] = true
		}
		roles["broker"] = broker
	}

	details["roles"] = roles

	return details
}

func WelcomeToCapnproto(w *messages.Welcome) ([]byte, error) {
	msg, seg, err := capnp.NewMessage(capnp.SingleSegment(nil))
	if err != nil {
		return nil, err
	}

	welcome, err := gen.NewRootWelcome(seg)
	if err != nil {
		return nil, err
	}

	welcome.SetSessionID(w.SessionID())

	authid, ok := w.Details()["authid"].(string)
	if ok {
		if err := welcome.SetAuthid(authid); err != nil {
			return nil, err
		}
	}

	authrole, ok := w.Details()["authrole"].(string)
	if ok {
		if err := welcome.SetAuthrole(authrole); err != nil {
			return nil, err
		}
	}

	authmethod, ok := w.Details()["authmethod"].(string)
	if ok {
		if err := welcome.SetAuthmethod(authmethod); err != nil {
			return nil, err
		}
	}

	authprovider, ok := w.Details()["authprovider"].(string)
	if ok {
		if err := welcome.SetAuthprovider(authprovider); err != nil {
			return nil, err
		}
	}

	roles, err := welcome.NewRoles()
	if err != nil {
		return nil, err
	}

	rolesMap, ok := w.Details()["roles"].(map[string]any)
	if ok {
		if dealerMap, ok := rolesMap["dealer"].(map[string]any); ok {
			dealer, err := roles.NewDealer()
			if err == nil {
				if v, ok := dealerMap["caller_identification"].(bool); ok {
					dealer.SetCallerIdentification(v)
				}
				if v, ok := dealerMap["call_timeout"].(bool); ok {
					dealer.SetCallTimeout(v)
				}
				if v, ok := dealerMap["call_canceling"].(bool); ok {
					dealer.SetCallCanceling(v)
				}
				if v, ok := dealerMap["progressive_call_results"].(bool); ok {
					dealer.SetProgressiveCallResults(v)
				}
				if v, ok := dealerMap["pattern_based_registration"].(bool); ok {
					dealer.SetPatternBasedRegistration(v)
				}
				_ = roles.SetDealer(dealer)
			}
		}

		if brokerMap, ok := rolesMap["broker"].(map[string]any); ok {
			broker, err := roles.NewBroker()
			if err == nil {
				if v, ok := brokerMap["publisher_identification"].(bool); ok {
					broker.SetPublisherIdentification(v)
				}
				if v, ok := brokerMap["publisher_exclusion"].(bool); ok {
					broker.SetPublisherExclusion(v)
				}
				if v, ok := brokerMap["acknowledge_event_received"].(bool); ok {
					broker.SetAcknowledgeEventReceived(v)
				}
				if v, ok := brokerMap["pattern_based_subscription"].(bool); ok {
					broker.SetPatternBasedSubscription(v)
				}
				_ = roles.SetBroker(broker)
			}
		}
	}

	if err := welcome.SetRoles(roles); err != nil {
		return nil, err
	}

	data, err := msg.MarshalPacked()
	if err != nil {
		return nil, err
	}

	return PrependHeader(messages.MessageTypeWelcome, data, nil), nil
}

func CapnprotoToWelcome(data []byte) (*messages.Welcome, error) {
	msg, err := capnp.UnmarshalPacked(data)
	if err != nil {
		return nil, err
	}

	welcome, err := gen.ReadRootWelcome(msg)
	if err != nil {
		return nil, err
	}

	return messages.NewWelcomeWithFields(NewWelcomeFields(&welcome)), nil
}

File: python/wamp_msgs_capnp/parsers/welcome.py
import os
from typing import Any
from pathlib import Path

import capnp
from wampproto.messages.welcome import Welcome, IWelcomeFields

from wamp_msgs_capnp.parsers import helpers

root_dir = Path(__file__).resolve().parents[1]
module_file = os.path.join(root_dir, "schemas", "welcome.capnp")
welcome_capnp = capnp.load(str(module_file))


class WelcomeFields(IWelcomeFields):
    def __init__(self, gen):
        self._gen = gen

    @property
    def session_id(self) -> int:
        return self._gen.sessionID

    @property
    def details(self) -> dict:
        details = {
            "authid": self._gen.authid,
            "authrole": self._gen.authrole,
            "authmethod": self._gen.authmethod,
            "authprovider": self._gen.authprovider,
        }

        if self._gen.roles.dealer.callerIdentification:
            details["dealer"]["caller_identification"] = True

        if self._gen.roles.dealer.callTimeout:
            details["dealer"]["call_timeout"] = True

        if self._gen.roles.dealer.callCanceling:
            details["dealer"]["call_canceling"] = True

        if self._gen.roles.dealer.progressiveCallResults:
            details["dealer"]["progressive_call_results"] = True

        if self._gen.roles.dealer.patternBasedRegistration:
            details["dealer"]["pattern_based_registration"] = True

        if self._gen.roles.dealer.sharedRegistration:
            details["dealer"]["shared_registration"] = True

        r = self._gen.roles.callee
        details["broker"] = {
            "broker_identification": r.publisherIdentification,
            "publisher_exclusion": r.publisherExclusion,
            "acknowledge_event_received": r.acknowledgeEventReceived,
            "pattern_based_subscription": r.patternBasedSubscription,
        }

        if self._gen.role.broker.publisherIdentification:
            details["broker"]["publisher_identification"] = True

        if self._gen.role.broker.publisherExclusion:
            details["broker"]["publisher_exclusion"] = True

        if self._gen.role.broker.acknowledgeEventReceived:
            details["broker"]["acknowledge_event_received"] = True

        if self._gen.role.broker.patternBasedSubscription:
            details["broker"]["pattern_based_subscription"] = True

        return details

    @property
    def roles(self) -> dict[str, Any]:
        return self._gen.roles

    @property
    def authid(self) -> str:
        return self._gen.authid

    @property
    def authrole(self) -> str:
        return self._gen.authrole

    @property
    def authmethod(self) -> str:
        return self._gen.authmethod

    @property
    def authextra(self) -> dict[str, Any]:
        return {}


def welcome_to_capnproto(w: Welcome) -> bytes:
    welcome = welcome_capnp.Welcome.new_message()

    welcome.sessionID = w.session_id
    welcome.authid = w.authid
    welcome.authrole = w.authrole
    welcome.authmethod = w.authmethod

    roles = welcome.init("roles")
    if "dealer" in w.roles:
        dealer = roles.init("dealer")
        dealer.callerIdentification = w.roles["dealer"].get("caller_identification", False)
        dealer.callTimeout = w.roles["dealer"].get("call_timeout", False)
        dealer.callCanceling = w.roles["dealer"].get("call_canceling", False)
        dealer.progressiveCallResults = w.roles["dealer"].get("progressive_call_results", False)
        dealer.patternBasedSubscription = w.roles["dealer"].get("pattern_based_registration", False)
        dealer.sharedRegistration = w.roles["dealer"].get("shared_registration", False)

    if "broker" in w.roles:
        broker = roles.init("broker")
        broker.publisherIdentification = w.roles["broker"].get("publisher_identification", False)
        broker.publisherExclusion = w.roles["broker"].get("publisher_exclusion", False)
        broker.acknowledgeEventReceived = w.roles["broker"].get("acknowledge_event_received", False)
        broker.patternBasedSubscription = w.roles["broker"].get("pattern_based_subscription", False)

    welcome.roles = roles

    packed_data = welcome.to_bytes_packed()

    return helpers.prepend_header(Welcome.TYPE, packed_data, b"")


def capnproto_to_welcome(data: bytes) -> Welcome:
    message_data, _ = helpers.extract_message(data)
    welcome_obj = welcome_capnp.Welcome.from_bytes_packed(message_data)

    return Welcome(WelcomeFields(welcome_obj))
